### PR TITLE
[STAL-982] Add basic logging

### PR DESCRIPTION
**Pull request by jasonforal**
# STAL-982

# Implement basic logging

## What problem are you trying to solve?
Give (limited) ability to manually review logs and inspect requests via logging.

## What is your solution?
1. **Use an event/subscriber architecture**
* Avoids making assumptions about how downstream consumers will consume events. For example, instead of forcing `println!("foo")` in [server](https://github.com/DataDog/datadog-static-analyzer/tree/main/server), we construct an event with `tracing::info!("foo")` and allow the user to handle this stream of tracing events however they want (e.g. print to stdout).

2. Use the `tracing` crate
* This crate is backed by the Tokio project and thus has very low risk of abandonment.
* This crate revolves around spans and events, which are natural abstractions we'll need to instrument the service with performance metrics and (actual) tracing.

3. **Implement unique HTTP request id generation**
* We want to be able to manually inspect logs, so we need to be able to track a request across its lifecycle.
* Rocket currently does not generate unique IDs for each request.
* We implement a UUID v4 generator for requests that don't send an `X-Request-Id` header, and we surface this UUID in the logs. This is most appropriately implemented using Rocket's Fairing API, as it's used for every request and doesn't modify control flow.

## Alternatives considered

* HTTP request id was chosen over a full-blown "trace id" for simplicity. It can be refactored later if necessary.
* `tracing` was used instead of Rust std `log` because we'll eventually want support for spans for metrics/tracing. Additionally, for the stdout logging use case, `tracing` provides the same abstraction benefits of the `log` facade.

## What the reviewer should know
1. The `DEBUG` log level will print the (validated) POST data for the `/analyze` and `/get-tree-sitter-ast` endpoints. We currently don't enforce a length limit beyond Rocket's max byte length, which could bloat server logs.
2. This PR uses `INFO` level events for malformed user input, preferring to use `WARN` for things in "our" control. (I'm not strongly opinionated on this, though).
3. The `X-Request-Id` header is echoed if supplied by the user. We don't currently treat this as untrusted user input (e.g. enforce a charset or length)

# Samples
```
2024-01-02T17:40:15.204675Z  WARN rocket::launch: 🚀 Rocket has launched from http://127.0.0.1:8000
2024-01-02T17:40:47.455938Z DEBUG http_request{http.request_id="dcd39036-a81f-48c5-b50e-af27c374f434" http.method="POST" http.uri="/get-treesitter-ast"}: datadog_static_analyzer_server::datadog_static_analyzer_server::endpoints: TreeSitterRequest { language: Csharp, file_encoding: "utf-8", code_base64: "dXNpbmcgU3lzdGVtLlhtbDsKCmNsYXNzIE15Q2xhc3MgewogICAgfk15Q2xhc3MoKQogICAgewogICAgICAgIGlmIChmb28pIHsKICAgICAgICAgICAgdGhyb3cgbmV3IE15RXhjZXB0aW9uKCk7CiAgICAgICAgfQogICAgICAgIAogICAgfQp9Cg==" }
2024-01-02T17:40:47.456018Z DEBUG http_request{http.request_id="dcd39036-a81f-48c5-b50e-af27c374f434" http.method="POST" http.uri="/get-treesitter-ast"}:process_tree_sitter_tree_request: server::tree_sitter_tree: Processing tree-sitter AST generation request
2024-01-02T17:40:47.456117Z DEBUG http_request{http.request_id="dcd39036-a81f-48c5-b50e-af27c374f434" http.method="POST" http.uri="/get-treesitter-ast"}:process_tree_sitter_tree_request: server::tree_sitter_tree: Getting tree-sitter tree (code length: 139 bytes)
2024-01-02T17:40:47.457160Z  INFO http_request{http.request_id="dcd39036-a81f-48c5-b50e-af27c374f434" http.method="POST" http.uri="/get-treesitter-ast"}:process_tree_sitter_tree_request: server::tree_sitter_tree: Successfully completed tree-sitter AST generation
2024-01-02T17:40:47.756124Z DEBUG http_request{http.request_id="0bec3647-79c4-4280-9ee3-ec6540030861" http.method="POST" http.uri="/analyze"}: datadog_static_analyzer_server::datadog_static_analyzer_server::endpoints: AnalysisRequest { filename: "NotCompliant2.cs", language: Csharp, file_encoding: "utf-8", code_base64: "dXNpbmcgU3lzdGVtLlhtbDsKCmNsYXNzIE15Q2xhc3MgewogICAgfk15Q2xhc3MoKQogICAgewogICAgICAgIGlmIChmb28pIHsKICAgICAgICAgICAgdGhyb3cgbmV3IE15RXhjZXB0aW9uKCk7CiAgICAgICAgfQogICAgICAgIAogICAgfQp9Cg==", rules: [ServerRule { name: "finalizer-no-exception", short_description_base64: None, description_base64: None, category: Some(BestPractices), severity: Some(Warning), language: Csharp, rule_type: TreeSitterQuery, entity_checked: None, code_base64: "LyoqCiAqIEEgZnVuY3Rpb24gZXhlY3V0ZWQgb3ZlciBlYWNoIGNhcHR1cmVkIEFTVCBub2RlCiAqIEBwYXJhbSB7UXVlcnl9IHF1ZXJ5IC0gQSB0cmVlIHNpdHRlciBxdWVyeS4KICogQHBhcmFtIHtzdHJpbmd9IGZpbGVuYW1lIC0gQSBmaWxlIG5hbWUuCiAqIEBwYXJhbSB7c3RyaW5nfSBjb2RlIC0gQSB0ZXN0IGNvZGUuCiAqIEByZXR1cm5zCiAqLwpmdW5jdGlvbiB2aXNpdChxdWVyeSwgZmlsZW5hbWUsIGNvZGUpIHsKICBjb25zdCB2ID0gcXVlcnkuY2FwdHVyZXNbInRocm93Il07CgoKICBpZih2KSB7CiAgICBjb25zdCBlID0gYnVpbGRFZGl0UmVtb3ZlKAogICAgICB2LnN0YXJ0LmxpbmUsIHYuc3RhcnQuY29sLAogICAgICB2LmVuZC5saW5lLCB2LmVuZC5jb2wKICAgICk7CiAgICBhZGRFcnJvcigKICAgICAgYnVpbGRFcnJvcigKICAgICAgICB2LnN0YXJ0LmxpbmUsIHYuc3RhcnQuY29sLAogICAgICAgIHYuZW5kLmxpbmUsIHYuZW5kLmNvbCwKICAgICAgICAiZG8gbm90IHVzZSB0aHJvdyBleGNlcHRpb24gaW4gZmluYWxpemVyIgogICAgICApLmFkZEZpeCgKICAgICAgICBidWlsZEZpeCgncmVtb3ZlJywgW2VdKQogICAgICApCiAgICApCiAgfQp9Cg==", checksum: Some("f77325a22884d09b74e1c45e94ee21a21a5fbc0417a5479890c91fb633e440fd"), pattern: None, tree_sitter_query_base64: Some("KGRlc3RydWN0b3JfZGVjbGFyYXRpb24KCiAgICBib2R5OiAoYmxvY2sKICAgICAgICBbCiAgICAgICAgICAgICh0aHJvd19zdGF0ZW1lbnQpIEB0aHJvdwogICAgICAgICAgICAoXwogICAgICAgICAgICAgICAgKHRocm93X3N0YXRlbWVudCkgQHRocm93CiAgICAgICAgICAgICkKICAgICAgICAgICAgKF8KICAgICAgICAgICAgICAgIChfCiAgICAgICAgICAgICAgICAgICAgKHRocm93X3N0YXRlbWVudCkgQHRocm93CiAgICAgICAgICAgICAgICApCiAgICAgICAgICAgICkKICAgICAgICBdCiAgICAgICAgCiAgICApCik="), variables: None }], options: Some(AnalysisRequestOptions { use_tree_sitter: Some(true), log_output: Some(true) }) }
2024-01-02T17:40:47.756559Z DEBUG http_request{http.request_id="0bec3647-79c4-4280-9ee3-ec6540030861" http.method="POST" http.uri="/analyze"}:process_analysis_request: server::request: Processing analysis request
2024-01-02T17:40:47.756684Z DEBUG http_request{http.request_id="c41feef2-7137-4dee-87de-3b45fdd9b37f" http.method="POST" http.uri="/analyze"}: datadog_static_analyzer_server::datadog_static_analyzer_server::endpoints: AnalysisRequest { filename: "NotCompliant.cs", language: Csharp, file_encoding: "utf-8", code_base64: "dXNpbmcgU3lzdGVtLlhtbDsKCmNsYXNzIE15Q2xhc3MgewogICAgfk15Q2xhc3MoKQogICAgewogICAgICAgIHRocm93IG5ldyBNeUV4Y2VwdGlvbigpOwogICAgfQp9Cg==", rules: [ServerRule { name: "finalizer-no-exception", short_description_base64: None, description_base64: None, category: Some(BestPractices), severity: Some(Warning), language: Csharp, rule_type: TreeSitterQuery, entity_checked: None, code_base64: "LyoqCiAqIEEgZnVuY3Rpb24gZXhlY3V0ZWQgb3ZlciBlYWNoIGNhcHR1cmVkIEFTVCBub2RlCiAqIEBwYXJhbSB7UXVlcnl9IHF1ZXJ5IC0gQSB0cmVlIHNpdHRlciBxdWVyeS4KICogQHBhcmFtIHtzdHJpbmd9IGZpbGVuYW1lIC0gQSBmaWxlIG5hbWUuCiAqIEBwYXJhbSB7c3RyaW5nfSBjb2RlIC0gQSB0ZXN0IGNvZGUuCiAqIEByZXR1cm5zCiAqLwpmdW5jdGlvbiB2aXNpdChxdWVyeSwgZmlsZW5hbWUsIGNvZGUpIHsKICBjb25zdCB2ID0gcXVlcnkuY2FwdHVyZXNbInRocm93Il07CgoKICBpZih2KSB7CiAgICBjb25zdCBlID0gYnVpbGRFZGl0UmVtb3ZlKAogICAgICB2LnN0YXJ0LmxpbmUsIHYuc3RhcnQuY29sLAogICAgICB2LmVuZC5saW5lLCB2LmVuZC5jb2wKICAgICk7CiAgICBhZGRFcnJvcigKICAgICAgYnVpbGRFcnJvcigKICAgICAgICB2LnN0YXJ0LmxpbmUsIHYuc3RhcnQuY29sLAogICAgICAgIHYuZW5kLmxpbmUsIHYuZW5kLmNvbCwKICAgICAgICAiZG8gbm90IHVzZSB0aHJvdyBleGNlcHRpb24gaW4gZmluYWxpemVyIgogICAgICApLmFkZEZpeCgKICAgICAgICBidWlsZEZpeCgncmVtb3ZlJywgW2VdKQogICAgICApCiAgICApCiAgfQp9Cg==", checksum: Some("f77325a22884d09b74e1c45e94ee21a21a5fbc0417a5479890c91fb633e440fd"), pattern: None, tree_sitter_query_base64: Some("KGRlc3RydWN0b3JfZGVjbGFyYXRpb24KCiAgICBib2R5OiAoYmxvY2sKICAgICAgICBbCiAgICAgICAgICAgICh0aHJvd19zdGF0ZW1lbnQpIEB0aHJvdwogICAgICAgICAgICAoXwogICAgICAgICAgICAgICAgKHRocm93X3N0YXRlbWVudCkgQHRocm93CiAgICAgICAgICAgICkKICAgICAgICAgICAgKF8KICAgICAgICAgICAgICAgIChfCiAgICAgICAgICAgICAgICAgICAgKHRocm93X3N0YXRlbWVudCkgQHRocm93CiAgICAgICAgICAgICAgICApCiAgICAgICAgICAgICkKICAgICAgICBdCiAgICAgICAgCiAgICApCik="), variables: None }], options: Some(AnalysisRequestOptions { use_tree_sitter: Some(true), log_output: Some(true) }) }
2024-01-02T17:40:47.757002Z DEBUG http_request{http.request_id="c41feef2-7137-4dee-87de-3b45fdd9b37f" http.method="POST" http.uri="/analyze"}:process_analysis_request: server::request: Processing analysis request
2024-01-02T17:40:47.812036Z  INFO http_request{http.request_id="c41feef2-7137-4dee-87de-3b45fdd9b37f" http.method="POST" http.uri="/analyze"}:process_analysis_request: server::request: Successfully completed analysis for 1 rule (finalizer-no-exception)
2024-01-02T17:40:47.812036Z  INFO http_request{http.request_id="0bec3647-79c4-4280-9ee3-ec6540030861" http.method="POST" http.uri="/analyze"}:process_analysis_request: server::request: Successfully completed analysis for 1 rule (finalizer-no-exception)
2024-01-02T17:40:47.812824Z DEBUG http_request{http.request_id="77c46c8d-e4b8-48e9-8715-37759e0c9dc3" http.method="POST" http.uri="/analyze"}: datadog_static_analyzer_server::datadog_static_analyzer_server::endpoints: AnalysisRequest { filename: "NotCompliant2.cs", language: Csharp, file_encoding: "utf-8", code_base64: "dXNpbmcgU3lzdGVtLlhtbDsKCmNsYXNzIE15Q2xhc3MgewogICAgfk15Q2xhc3MoKQogICAgewogICAgICAgIGlmIChmb28pIHsKICAgICAgICAgICAgdGhyb3cgbmV3IE15RXhjZXB0aW9uKCk7CiAgICAgICAgfQogICAgICAgIAogICAgfQp9Cg==", rules: [ServerRule { name: "finalizer-no-exception", short_description_base64: None, description_base64: None, category: Some(BestPractices), severity: Some(Warning), language: Csharp, rule_type: TreeSitterQuery, entity_checked: None, code_base64: "LyoqCiAqIEEgZnVuY3Rpb24gZXhlY3V0ZWQgb3ZlciBlYWNoIGNhcHR1cmVkIEFTVCBub2RlCiAqIEBwYXJhbSB7UXVlcnl9IHF1ZXJ5IC0gQSB0cmVlIHNpdHRlciBxdWVyeS4KICogQHBhcmFtIHtzdHJpbmd9IGZpbGVuYW1lIC0gQSBmaWxlIG5hbWUuCiAqIEBwYXJhbSB7c3RyaW5nfSBjb2RlIC0gQSB0ZXN0IGNvZGUuCiAqIEByZXR1cm5zCiAqLwpmdW5jdGlvbiB2aXNpdChxdWVyeSwgZmlsZW5hbWUsIGNvZGUpIHsKICBjb25zdCB2ID0gcXVlcnkuY2FwdHVyZXNbInRocm93Il07CgoKICBpZih2KSB7CiAgICBjb25zdCBlID0gYnVpbGRFZGl0UmVtb3ZlKAogICAgICB2LnN0YXJ0LmxpbmUsIHYuc3RhcnQuY29sLAogICAgICB2LmVuZC5saW5lLCB2LmVuZC5jb2wKICAgICk7CiAgICBhZGRFcnJvcigKICAgICAgYnVpbGRFcnJvcigKICAgICAgICB2LnN0YXJ0LmxpbmUsIHYuc3RhcnQuY29sLAogICAgICAgIHYuZW5kLmxpbmUsIHYuZW5kLmNvbCwKICAgICAgICAiZG8gbm90IHVzZSB0aHJvdyBleGNlcHRpb24gaW4gZmluYWxpemVyIgogICAgICApLmFkZEZpeCgKICAgICAgICBidWlsZEZpeCgncmVtb3ZlJywgW2VdKQogICAgICApCiAgICApCiAgfQp9Cg==", checksum: Some("f77325a22884d09b74e1c45e94ee21a21a5fbc0417a5479890c91fb633e440fd"), pattern: None, tree_sitter_query_base64: Some("KGRlc3RydWN0b3JfZGVjbGFyYXRpb24KCiAgICBib2R5OiAoYmxvY2sKICAgICAgICBbCiAgICAgICAgICAgICh0aHJvd19zdGF0ZW1lbnQpIEB0aHJvdwogICAgICAgICAgICAoXwogICAgICAgICAgICAgICAgKHRocm93X3N0YXRlbWVudCkgQHRocm93CiAgICAgICAgICAgICkKICAgICAgICAgICAgKF8KICAgICAgICAgICAgICAgIChfCiAgICAgICAgICAgICAgICAgICAgKHRocm93X3N0YXRlbWVudCkgQHRocm93CiAgICAgICAgICAgICAgICApCiAgICAgICAgICAgICkKICAgICAgICBdCiAgICAgICAgCiAgICApCik="), variables: None }], options: Some(AnalysisRequestOptions { use_tree_sitter: Some(true), log_output: Some(true) }) }
2024-01-02T17:40:47.812951Z DEBUG http_request{http.request_id="77c46c8d-e4b8-48e9-8715-37759e0c9dc3" http.method="POST" http.uri="/analyze"}:process_analysis_request: server::request: Processing analysis request
2024-01-02T17:40:47.849368Z  INFO http_request{http.request_id="77c46c8d-e4b8-48e9-8715-37759e0c9dc3" http.method="POST" http.uri="/analyze"}:process_analysis_request: server::request: Successfully completed analysis for 1 rule (finalizer-no-exception)
```

### CODE REVIEW

File: server/src/request.rs [hunk 1]
```diff
@@ -65,52 +92,60 @@ pub fn process_analysis_request(request: AnalysisRequest) -> AnalysisResponse {
...
 
+    let rules_count = rules.len();
+    let rules_str = if rules_count == 1 { "rule" } else { "rules" };
+    let rules_list = rules
+        .iter()
+        .map(|r| r.name.as_str())
+        .collect::<Vec<&str>>()
+        .join(", ");
     // execute the rule. If we fail to convert, return an error.
-    match rules_converted {
```
Comment (dastrong) [inline server/src/request.rs:77]: Are we sure that we don't need this match here anymore?

File: server/src/request.rs [hunk 2]
```diff
@@ -65,52 +92,60 @@ pub fn process_analysis_request(request: AnalysisRequest) -> AnalysisResponse {
...
 
-            AnalysisResponse {
-                rule_responses,
-                errors: vec![],
-            }
-        }
-        Err(_) => AnalysisResponse {
-            rule_responses: vec![],
-            errors: vec![ERROR_DECODING_BASE64.to_string()],
-        },
```
Comment (dastrong) [inline server/src/request.rs:113]: Related to the comment above, but is this error handler needed

File: bins/src/bin/datadog_static_analyzer_server/endpoints.rs [hunk 3]
```diff
@@ -101,12 +103,16 @@ fn languages() -> Value {
 }
 
 #[rocket::post("/analyze", format = "application/json", data = "<request>")]
-fn analyze(request: Json<AnalysisRequest>) -> Value {
+fn analyze(span: TraceSpan, request: Json<AnalysisRequest>) -> Value {
+    let _entered = span.enter();
+    tracing::debug!("{:?}", &request.0);
```
Comment (dastrong) [inline bins/src/bin/datadog_static_analyzer_server/endpoints.rs:108]: Just for my curiosity, what does `request.0` render? Did you have an example?

**Review by dastrong — APPROVED** [review]
For logs, I believe we're trying to enforce prepending the log with an `[]` to make it easier to distinguish logs easier, but that's just a nit here.
ex) `[example]`

File: server/src/request.rs (hunk 1 above)
Comment (juli1) [reply server/src/request.rs:77]: This is being addressed on line 74.

File: server/src/request.rs (hunk 2 above)
Comment (juli1) [reply server/src/request.rs:113]: line 74

**Review by juli1 — CHANGES_REQUESTED** [review]

**Review by juli1 — APPROVED** [review]

**Pull request merged**